reference/shaders-msl-no-opt/asm/frag/texture-access.swizzle.asm.frag

```diff
@@ -7,6 +7,7 @@ using namespace metal;
 
 struct spvAux
 {
+    uint vertexCount;
```

Comment:
I'm a bit concerned about this magic AuxBuffer. I suppose, we'll need to make sure MoltenVK stays in sync with these changes? @billhollings, is there any integration testing in place to catch changes like these when you pull in latest SPIRV-Cross?

Reply:
@HansKristian-Work You raise a good point...no...there is no explicit API integration testing at the buffer-contents level.

@cdavis5e What do you think? I expect `spvAux` will evolve slowly...and that we would make changes to both SPIRV-Cross and MoltenVK simultaneously...so I am not **_too_** concerned about this right now. But in the larger scope..where SPIRV-Cross MSL is influenced by more than MoltenVK...there is the possibility for silent surprises.

One option I can think of is to enable changes to the `spvAux` struct through API Option switches. Each member of `spvAux` could have a corresponding API option bool (eg. an `CompilerMSL::Options::spvAuxFeatures` which looks the same as `spvAux`but with all bools)...and SPIRV-Cross would only emit each member of `spvAux` if the parallel Options struct member enabled it.

But would that grow to be a huge PITA?

Another option would be to create a really simple test GLSL or SPIR-V shader file within the MoltenVK repo that uses the aux buffer...and run `test-shaders.py` on it from the MoltenVK `fetchDependencies` script. That would identify that the struct has changed when a new version of SPIRV-Cross is brought into MoltenVK.

Reply:
@cdavis5e Regarding my last option...a test shader within MoltenVK...if you can provide me with a simple GLSL shader and resulting MSL reference shader that uses `spvAux`...I can put them into a new `spirv-test` directory within MoltenVK and modify `fetchDependencies` to run `test-shaders.py` on that directory whenever SPIRV-Cross is updated.

We can then add additional shaders as needed when the SPIRV-Cross API changes in an similarly hidden ways.

Reply:
I think some kind of query would make sense so that the consumer of SPIRV-Cross can know the ABI layout of this structure. Happy to support that, but I'd rather the consumers of this API propose it. From SPIRV-Cross' point of view, this should be as simple as bumping a few constants whenever the AuxBuffer layout changes, it's the downstream which has to deal with it.

Reply:
> What do you think? I expect `spvAux` will evolve slowly...and that we would make changes to both SPIRV-Cross and MoltenVK simultaneously

As do I.

> One option I can think of is to enable changes to the `spvAux` struct through API Option switches.
> But would that grow to be a huge PITA?

I don't expect this to have too many more members. The only one I can think of off the top of my head right now that I haven't added yet is `BuiltInDrawIndex`. In the past, I mentioned `BuiltInViewIndex`, but I'm considering deriving that from the `BuiltInInstanceIndex`; the way `VK_KHR_multiview` is specified implies this is an acceptable way to implement this.

> if you can provide me with a simple GLSL shader and resulting MSL reference shader that uses `spvAux`...I can put them into a new `spirv-test` directory within MoltenVK and modify `fetchDependencies` to run `test_shaders.py` on that directory whenever SPIRV-Cross is updated.

The GLSL shader:
```glsl
#version 450

layout(binding = 0) uniform sampler2D t0;

layout(location = 0) out vec4 FragColor;

void main()
{
    FragColor = texture(t0, gl_FragCoord.xy);
}
```

Compiled with:
```console
$ glslangValidator -V --target-env vulkan1.1 spv-aux-buffer.frag -o spv-aux-buffer.spv
$ spirv-cross --entry main --output spv-aux-buffer.frag.metal spv-aux-buffer.spv --msl --msl-version 10200 --msl-swizzle-texture-samples
```

The Metal shader:
```c++
#pragma clang diagnostic ignored "-Wmissing-prototypes"

#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct spvAux
{
    uint vertexCount;
    uint swizzleConst[1];
};

struct main0_out
{
    float4 FragColor [[color(0)]];
};

enum class spvSwizzle : uint
{
    none = 0,
    zero,
    one,
    red,
    green,
    blue,
    alpha
};

template<typename T> struct spvRemoveReference { typedef T type; };
template<typename T> struct spvRemoveReference<thread T&> { typedef T type; };
template<typename T> struct spvRemoveReference<thread T&&> { typedef T type; };
template<typename T> inline constexpr thread T&& spvForward(thread typename spvRemoveReference<T>::type& x)
{
    return static_cast<thread T&&>(x);
}
template<typename T> inline constexpr thread T&& spvForward(thread typename spvRemoveReference<T>::type&& x)
{
    return static_cast<thread T&&>(x);
}

template<typename T>
inline T spvGetSwizzle(vec<T, 4> x, T c, spvSwizzle s)
{
    switch (s)
    {
        case spvSwizzle::none:
            return c;
        case spvSwizzle::zero:
            return 0;
        case spvSwizzle::one:
            return 1;
        case spvSwizzle::red:
            return x.r;
        case spvSwizzle::green:
            return x.g;
        case spvSwizzle::blue:
            return x.b;
        case spvSwizzle::alpha:
            return x.a;
    }
}

// Wrapper function that swizzles texture samples and fetches.
template<typename T>
inline vec<T, 4> spvTextureSwizzle(vec<T, 4> x, uint s)
{
    if (!s)
        return x;
    return vec<T, 4>(spvGetSwizzle(x, x.r, spvSwizzle((s >> 0) & 0xFF)), spvGetSwizzle(x, x.g, spvSwizzle((s >> 8) & 0xFF)), spvGetSwizzle(x, x.b, spvSwizzle((s >> 16) & 0xFF)), spvGetSwizzle(x, x.a, spvSwizzle((s >> 24) & 0xFF)));
}

template<typename T>
inline T spvTextureSwizzle(T x, uint s)
{
    return spvTextureSwizzle(vec<T, 4>(x, 0, 0, 1), s).x;
}

// Wrapper function that swizzles texture gathers.
template<typename T, typename Tex, typename... Ts>
inline vec<T, 4> spvGatherSwizzle(sampler s, const thread Tex& t, Ts... params, component c, uint sw) METAL_CONST_ARG(c)
{
    if (sw)
    {
        switch (spvSwizzle((sw >> (uint(c) * 8)) & 0xFF))
        {
            case spvSwizzle::none:
                break;
            case spvSwizzle::zero:
                return vec<T, 4>(0, 0, 0, 0);
            case spvSwizzle::one:
                return vec<T, 4>(1, 1, 1, 1);
            case spvSwizzle::red:
                return t.gather(s, spvForward<Ts>(params)..., component::x);
            case spvSwizzle::green:
                return t.gather(s, spvForward<Ts>(params)..., component::y);
            case spvSwizzle::blue:
                return t.gather(s, spvForward<Ts>(params)..., component::z);
            case spvSwizzle::alpha:
                return t.gather(s, spvForward<Ts>(params)..., component::w);
        }
    }
    switch (c)
    {
        case component::x:
            return t.gather(s, spvForward<Ts>(params)..., component::x);
        case component::y:
            return t.gather(s, spvForward<Ts>(params)..., component::y);
        case component::z:
            return t.gather(s, spvForward<Ts>(params)..., component::z);
        case component::w:
            return t.gather(s, spvForward<Ts>(params)..., component::w);
    }
}

// Wrapper function that swizzles depth texture gathers.
template<typename T, typename Tex, typename... Ts>
inline vec<T, 4> spvGatherCompareSwizzle(sampler s, const thread Tex& t, Ts... params, uint sw) 
{
    if (sw)
    {
        switch (spvSwizzle(sw & 0xFF))
        {
            case spvSwizzle::none:
            case spvSwizzle::red:
                break;
            case spvSwizzle::zero:
            case spvSwizzle::green:
            case spvSwizzle::blue:
            case spvSwizzle::alpha:
                return vec<T, 4>(0, 0, 0, 0);
            case spvSwizzle::one:
                return vec<T, 4>(1, 1, 1, 1);
        }
    }
    return t.gather_compare(s, spvForward<Ts>(params)...);
}

fragment main0_out main0(constant spvAux& spvAuxBuffer [[buffer(30)]], texture2d<float> t0 [[texture(0)]], sampler t0Smplr [[sampler(0)]], float4 gl_FragCoord [[position]])
{
    main0_out out = {};
    constant uint32_t& t0Swzl = spvAuxBuffer.swizzleConst[0];
    out.FragColor = spvTextureSwizzle(t0.sample(t0Smplr, gl_FragCoord.xy), t0Swzl);
    return out;
}
```

Reply:
If we're changing AuxBuffer, could we just add reserved fields for all the relevant driver data which may or may not be needed going forward? Avoiding churn on the ABI versions sounds like a good idea to me.

Reply:
AuxBuffer should also only contain data which is known at CPU recording time, FWIW. How would DrawIndex be supported? Unrolled on CPU side?

Reply:
> AuxBuffer should also only contain data which is known at CPU recording time, FWIW. How would DrawIndex be supported? Unrolled on CPU side?

That's how multi draws are already implemented in MoltenVK, precisely because Metal doesn't really support them directly.

Reply:
I have added a structure to hold optional members of the aux buffer.

Reply:
On the subject of tracking structural changes to `spvAux`...let's just take a KISS approach and define a macro in `spriv_msl.hpp`:

    #define SPIRV_CROSS_MSL_AUX_BUFFER_STRUCT_VERSION  2

which gets incremented whenever `spvAux` changes.

We can then include similar `XXX_SUPPORTED_MSL_AUX_BUFFER_STRUCT_VERSION` macros in MoltenVK & gfx, and test them during builds.

Reply:
Agreed. Done.

Reply:
@cdavis5e Sorry...it looks like our messages crossed in cyber-space. On reflection I think my initial suggestion of a feature struct was unnecessarily complicated.